### PR TITLE
perf(codec): optimize encode functions for modern CPUs

### PR DESCRIPTION
**Pull request by larseggert**
(This is me vibing with Claude again.)

### CODE REVIEW

**Comment by larseggert** [comment]
So Claude hallucinated checking the assembly with godbolt.org :-) Second attempt with more explicit instructions.

**Comment by larseggert** [comment]
Is this worth it?

File: neqo-common/src/codec.rs [hunk 1]
```diff
@@ -272,30 +272,33 @@ impl<B: Buffer> Encoder<B> {
     /// # Panics
     ///
     /// When `n` is outside the range `1..=8`.
+    #[inline]
```
Comment (mxinden) [inline neqo-common/src/codec.rs]: Unless we have solid  proof, my rule of thumb for `#[inline]` is, that the compiler knows better.

File: neqo-common/src/codec.rs [hunk 2]
```diff
@@ -272,30 +272,33 @@ impl<B: Buffer> Encoder<B> {
...
         let v = v.into();
         assert!(n > 0 && n <= 8);
-        for i in 0..n {
-            self.encode_byte(((v >> (8 * (n - i - 1))) & 0xff) as u8);
-        }
-        self
+        // Using to_be_bytes() generates better assembly, especially for odd sizes
+        // where it uses rev + strh instead of multiple strb instructions
+        let bytes = v.to_be_bytes();
+        self.encode(&bytes[8 - n..])
```
Comment (mxinden) [inline neqo-common/src/codec.rs:280]: If not for the sake of performance, I think we should do this for the sake of simplicity.

**Comment by larseggert** [comment]
> prepend a commit

Sorry, not sure what you mean here?

**Comment by mxinden** [comment]
Sorry for the brevity. Let me expand.

- `cargo mutant` is able to alter the code in significant ways without test failures.
- Thus I assume this code isn't well tested.
- Given that this pull request does substantial changes to the logic, it might introduce a bug, e.g. in the edge cases.
- Thus I suggest first writing a commit increasing test coverage, making sure it succeeds on `main`.
- Then running the new tests on this patch to make sure it still succeeds.

**Comment by larseggert** [comment]
So I think the `mutants` check here is meaningless, because for the code in question
```rust
            () if v < (1 << 6) => self.encode_uint(1, v),
            () if v < (1 << 14) => self.encode_uint(2, v | (1 << 14)),
            () if v < (1 << 30) => self.encode_uint(4, v | (2 << 30)),
            () if v < (1 << 62) => self.encode_uint(8, v | (3 << 62)),
```
or (`|`) and xor (`^`) produce the same result. We're taking a value smaller than some boundary, and or-ing or xor-ing it with a value larger than the boundary. Same result.

**Review by martinthomson — APPROVED** [review]
